### PR TITLE
nclient4: Ensure server identifier matches

### PR DESCRIPTION
**Pull request by mfischer**
Ensure server identifier matches the offer server identifier when
looking for ACKs or NAKs.

This protects against rogue DHCP servers NAKing requests the shouldn't.

Signed-off-by: Moritz Fischer <moritzf@google.com>

### CODE REVIEW

File: dhcpv4/nclient4/client.go [hunk 1]
```diff
@@ -482,7 +501,9 @@ func (c *Client) RequestFromOffer(ctx context.Context, offer *dhcpv4.DHCPv4, mod
 		return nil, fmt.Errorf("unable to create a request: %w", err)
 	}
 
-	response, err := c.SendAndRead(ctx, c.serverAddr, request, IsMessageType(dhcpv4.MessageTypeAck, dhcpv4.MessageTypeNak))
+	response, err := c.SendAndRead(ctx, c.serverAddr, request, IsAll(
+		IsCorrectServer(offer.ServerIdentifier),
```
Comment (hugelgupf) [inline dhcpv4/nclient4/client.go:505]: offer.ServerIdentifier()?

Comment (hugelgupf) [reply dhcpv4/nclient4/client.go:505]: I was actually just looking for a mention in the RFC for validating this field. It doesn't require it. I'd add a comment here saying something like

// Servers are supposed to only respond to Requests containing their server identifier, but sometimes non-compliant servers respond anyway. Clients are not required to validate this field, but servers are required to include the server identifier in their Offer per RFC 2131 Section 4.3.1 Table 3.

Comment (mfischer) [reply dhcpv4/nclient4/client.go]: Yeah, makes sense :) Unfortunately I know at least one misbehaving server :-)

**Review by hugelgupf — APPROVED** [review]

**Comment by hugelgupf** [comment]
i'm gonna wait till monday to see if @insomniacslk or @pmazzini have comments

**Review by pmazzini — APPROVED** [review]
lgtm

**Comment by hugelgupf** [comment]
lol it doesn't build, but travis didn't catch it? looking into it.

**Pull request merged**